### PR TITLE
feat: embassy-boot for rp2040

### PR DESCRIPTION
**Pull request by lulf**
Add embassy-boot support for RP2040, with examples for the Raspberry Pi Pico.

### CODE REVIEW

**Comment by lulf** [comment]
@danbev FYI ^^

**Comment by MathiasKoch** [comment]
Lol.. didn't see this before my PR landed.. Doh!

**Comment by MathiasKoch** [comment]
`PAGE_SIZE` & `WRITE_SIZE` shouldn't be needed as generics on the bootloader for rp, as those are given by the implementation, and re-exported by `embassy-rp`

**Comment by lulf** [comment]
@MathiasKoch Ah, I wasn't aware you had this in the works either lol. I have verified that the example works on the rpi pico. I'm fine with whatever PR being merged really, I have done some doc updates here as well, and I'll fix the page and write size.

**Comment by MathiasKoch** [comment]
I think if you change the page and write size, your PR looks more complete than mine :+1: 

I really just wanted to upstream this, to reduce the amount of "custom" code we have to maintain, now that I had a minute to do it :+1: 

Will just look through your implementation, and close my PR in favor of this. Great work! :+1:

File: examples/boot/application/rp/memory.x [hunk 1]
```diff
@@ -0,0 +1,15 @@
+MEMORY
+{
+  /* NOTE 1 K = 1 KiBi = 1024 bytes */
+  BOOT2                             : ORIGIN = 0x10000000, LENGTH = 0x100
```
Comment (MathiasKoch) [inline examples/boot/application/rp/memory.x:4]: Why have boot2 defined for the application?

File: examples/boot/bootloader/rp/memory.x [hunk 2]
```diff
@@ -0,0 +1,19 @@
+MEMORY
+{
+  /* NOTE 1 K = 1 KiBi = 1024 bytes */
+  BOOT2                             : ORIGIN = 0x10000000, LENGTH = 0x100
+  FLASH                             : ORIGIN = 0x10000100, LENGTH = 24K
+  BOOTLOADER_STATE                  : ORIGIN = 0x10007000, LENGTH = 4K
```
Comment (MathiasKoch) [inline examples/boot/bootloader/rp/memory.x:6]: why `0x10007000` with a bootloader of `24K`? shouldn't `0x10006000` suffice, if you do
`FLASH                             : ORIGIN = 0x10000100, LENGTH = 24K - 0x100`

**Review by MathiasKoch — APPROVED** [review]

File: examples/boot/application/rp/memory.x (hunk 1 above)
Comment (lulf) [reply examples/boot/application/rp/memory.x:4]: It's used in the linker env settings below, and I think it's great for documentation. Yes you can avoid it, but I think for an example it's useful to leave as much information in there as possible.

File: examples/boot/bootloader/rp/memory.x (hunk 2 above)
Comment (lulf) [reply examples/boot/bootloader/rp/memory.x]: Agreed!

**Review by lulf — COMMENTED** [review]
I've addressed the comments and pushed an update.

**Pull request merged**